atef/tools.py

```diff
@@ -0,0 +1,200 @@
+from __future__ import annotations
+
+import asyncio
+import dataclasses
+import re
+import shutil
+import sys
+import typing
+from dataclasses import field
+from typing import Any, ClassVar, Dict, List, Type, TypedDict, TypeVar, Union
+
+from . import serialization
+from .check import Result, Severity
+
+#: Arguments that can be passed to tools.
+ToolArguments = Dict[str, Any]
+T = TypeVar("T", bound="Tool")
+
+
+class ToolResult(TypedDict):
```

Comment:
My logic for switching to `TypedDict` here instead of a `dataclass` for each tool was along the lines of data interchange:
* External tools could more easily return JSON blobs that get deserialized into Python dictionaries
* Maybe such tools could also optionally take JSON as input via stdin or less preferably argv

Thinking a bit further - there's no reason we couldn't use `apischema.(de)serialize` here to take a dataclass and serialize it to JSON easily.

I think the argument for `TypedDict` is a bit weak here, so I'd be OK switching this back.

Reply:
I think it's a bit strange that this is the only TypedDict in the sea of atef dataclasses. But it's hard for me to have a bigger opinion about which is better to use since they are extremely similar result-wise.

Reply:
> I think it's a bit strange that this is the only TypedDict in the sea of atef dataclasses. 

Your comment plus my lack of good reason for it above is enough to say it was a mistake to go this route.  I'm switching it back to a dataclass. 

The `get_result_value_by_key` is flexible enough that it works with mappings/objects, so the conversion is very easy now.